metaflow/plugins/aws/eks/kubernetes_client.py

```diff
@@ -534,38 +583,50 @@ def status(self):
             return msg
         return "Job:Unknown"
 
-    @property
-    def has_succeeded(self):
+    def _check_has_succeeded(self):
         # Job is in a terminal state and the status is marked as succeeded
-        return self.is_done and bool(self._job["status"].get("succeeded"))
+        return self._check_is_job_done() and bool(self._job["status"].get("succeeded"))
 
-    @property
-    def has_failed(self):
+    def _check_has_failed(self):
         # Job is in a terminal state and either the status is marked as failed
         # or the Job is not allowed to launch any more pods
-        return self.is_done and (
+        return self._check_is_job_done() and (
             bool(self._job["status"].get("failed"))
             or (self._job["spec"]["parallelism"] == 0)
         )
 
-    @property
-    def is_running(self):
-        # The container is running. This happens when the Pod's phase is running
-        if not self.is_done:
+    def _check_is_running(self):
+        # Returns true if the container is running. There are two situations
+        # where is_running may return False, either:
+        # - the job is done
+        # - the container hasn't started *yet*
+        if not self._check_is_job_done():
             # If not done, check if pod has been assigned and is in Running
             # phase
             self._pod = self._fetch_pod()
-            return self._pod.get("status", {}).get("phase") == "Running"
+            pod_phase = self._pod.get("status", {}).get("phase")
+            if pod_phase == "Running":
+                return True
+            elif pod_phase in ("Succeeded", "Failed"):
```

Comment:
Maybe, I am missing something - why do we need to wait for 40 seconds for _us_ to mark the job as not running and not immediately?

Reply:
It is about the issue 3 in the PR description, and trying to make `RunningJob` state machine simpler.

The core issue here was that we'd rely on pod status to determine "running" and V1Job status to determine "done", and it is technically possible for pod to be "Succeeded" but V1Job would still have "active= 1" for a brief period, before the job controller updates the V1Job metadata. That caused a successful "RunningJob" go through these states:

1. not-running, not-done (haven't started yet)
2. running, not-done
3. not-running, not-done (again!)
4. not-running, done

I wanted to get rid of this 3rd state because it is confusing for the caller, sometimes doesn't even happen, and in this case practically this complexity caused this bug. For the caller, it is also indistinguishable from the very first state, which feels like a trap for future us. So in this PR, I've changed it so that if the pod is done but the job isn't, `is_running` just kinda wait it out so that next call to `is_done` is guaranteed to return True. Or in other words, after the job has started, it is impossible for it to be not-running and not-done at the same time.

Reply:
I agree, but we can return `False` for `_check_is_running` when `pod_phase in ("Succeeded", "Failed")` immediately and in `_check_is_job_done` we can add a check to for `_check_is_running` if the V1Job status is active. That should remove the need for a 40-second poll.

Reply:
So something like
```
def _check_is_job_done(..)
    if self._job["active"] and not self._check_is_running():
       return True
```
Effectively "folding" that intermediate state into "done" state instead?

That's an option, I think what's stopped me is that it would complicate `is_succeeded`/`is_failed` since they wouldn't be able to rely on V1Job status, as it may not be yet up to date.

Reply:
We can change the logic of `is_succeeded/is_failed` to also inspect the status of "pod_phase". We do have this issue where the "pod" itself may disappear depending on how the job was canceled.

Reply:
This will get us out of the business of maintaining the magic number of 40

Reply:
Hmm that may work but i think i need to draw the entire state diagram to make sure we're not missing anything. Btw remind me why don't we define `is_running()` as `V1JobStatus.active > 0` in the first place?

Reply:
`is_running` returns `True` only when the container starts running (V1Job may be active but for one reason or the other either the pod or the container may not have come up) - otherwise, we mark the job as waiting.

Reply:
Another point that we need to think through if we are introducing a check for `succeeded` and `failed` pod phases is what happens when a new pod phase is introduced. We can consider inspecting the PodConditions to have a greater degree of certainty, although it makes reasoning about the code a bit more convoluted. That was one of the reasons why I was only looking at the `Running` pod phase in the original implementation.

Reply:
Soo at the end of the day,
```
RunningJob.running := pod.phase is "Running"
RunningJob.done := V1JobStatus.succeeded > 0 OR V1JobStatus.failed > 0 OR pod.phase in ("Succeded", "Failed") OR V1JobSpec.parallelism is 0
RunningJob.succeeded := done AND (pod.phase is "Succeeded" OR V1JobStatus.succeeded > 0)
RunningJob.failed := done AND (pod.phase is "Failed" OR job parallelism is 0 OR V1JobStatus.failed > 0)
RunningJob.waiting = NOT (running OR done)
```
(if pod doesn't exist, pod.phase is assumed to be "null" above)

This should ensure that the job goes through `waiting -> running -> done` without weird intermediate states, and that a "done" job is always either "succeeded" or "failed".

Reply:
Seems reasonable to me (need to dust up my notes to find edge cases). This state diagram is pretty much the original implementation with accommodation for eventual consistency of job status? This still doesn't technically ensure that the job goes through `waiting -> running -> done` since pod phases can flip unpredictably to `unknown`. 

Also, as an aside, the implementation currently also tries to optimize for the number of pod and job status API calls, so the order in which these conditions are evaluated is important.

Reply:
Btw I have also removed another wait-loop where I'd wait for pod object to appear after the job was created, I think it is not necessary as long as we carefully handle pod-does-not-exist in the state machine logic